track/gpsd.py

```diff
@@ -0,0 +1,185 @@
+
+    def __enter__(self):
+        """Support usage of this class in 'with' statements."""
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """Support usage of this class in 'with' statements."""
+        self.close()
+
+    def close(self):
+        self.client.close()
+
+    # this is the method that will be called by alignment code
+    # requirements for successful execution:
+    # - don't time out
+    # - get a 3D fix
+    # - all relevant errors need to be within the specified bounds
+    # - we don't care about DGPS
+    # on success: returns a GPSLocation tuple
+    # on failure: raises GetLocationFailure with flags showing which requirements were not met
+    # parameters:
+    # - timeout: how much time to spend attempting to get a good fix before giving up
+    #            (set to None for no timeout, i.e. try forever)
+    # - need_3d: whether a 3D fix should be considered necessary
+    # - err_max: an Errors tuple containing the thresholds that each 95% error value must be below
+    #            for the fix to be considered satisfactory
+    #            (individual members of the tuple may be set to inf to ignore those ones)
+    # TODO: we *may* (not sure!) need to ensure that we get N consecutive reports that all meet our
+    #       requirements, before we declare ourselves successful and victorious
+    # TODO: determine whether we should enforce alt, and in which circumstances
+    #       (e.g. should we only enforce it if need_3d is True?)
+    #       enforcement would mean:
+    #       - change INIT_ALT from 0.0 to nan
+    #       - if isnan(self.location.alt) in _satisfies_criteria: return False
+    #       - if isnan(self.location.alt) in _raise_failure: add flag NO_ALT
+    #       also: if we are doing strictly a 2D fix, should we force self.location.alt to 0.0?
+    #       and do we even need to do that, or will the gps hw/sw do that for us...?
+    def get_location(self, timeout, need_3d, err_max):
+        if timeout is not None:
+            assert timeout > 0.0
+            t_start = perf_counter()
+
+        if need_3d:
+            fix_ok = lambda fix: fix == GPSFixType.FIX_3D
+        else:
+            fix_ok = lambda fix: fix == GPSFixType.FIX_3D or fix == GPSFixType.FIX_2D
+
+        for report in self.client:
+            if report['class'] != 'TPV': continue
+
+            if 'mode' in report:
+                if report.mode != 0: self.fix_type = GPSFixType(report.mode)
+                else:                self.fix_type = GPSFixType.NO_FIX
+
+            new_location = OrderedDict()
+            new_location['lat'] = report.lat if 'lat' in report else self.INIT_LOC['lat']
+            new_location['lon'] = report.lon if 'lon' in report else self.INIT_LOC['lon']
+            new_location['alt'] = report.alt if 'alt' in report else self.INIT_LOC['alt']
+            self.location = GPSLocation(**new_location) # TODO: make sure this works properly
```

Comment:
Why go through the extra step of making the OrderedDict first? Couldn't you do:
```
new_location = GPSLocation(
    lat=report.lat if 'lat' in report else self.INIT_LOC['lat'],
    lon=report.lon if 'lon' in report else self.INIT_LOC['lon'],
    alt=report.alt if 'alt' in report else self.INIT_LOC['alt'],
)
```
?

Reply:
Oh. Yeah. I actually didn't think of constructing it this way for whatever reason.

I *was* overly conscious about making a bunch of immutable copies of `namedtuple` if I were to "update things in place"; so I went for the `OrderedDict` approach. But this definitely makes more sense than that. :+1:

Reply:
Actually, I think I know why I did it that way originally... it made slightly more sense in a previous version of this code, when I was re-using the values already in `self.location` if there was no new value in the report. But I decided that that would allow for stale-and-wrong values to persist even if the gps actually had lost them so I went back to just setting them to "bad"/initial values whenever the report omits them.